src/core/tools/readFileTool.ts

```diff
@@ -435,6 +436,21 @@ export async function readFileTool(
 			try {
 				const [totalLines, isBinary] = await Promise.all([countFileLines(fullPath), isBinaryFile(fullPath)])
 
+				// Preemptive file size validation to prevent context overflow
+				const validation = await validateFileSizeForContext(fullPath, totalLines, maxReadFileLine, cline)
+				let effectiveMaxReadFileLine = maxReadFileLine
+				let validationNotice = ""
+
+				if (validation.shouldLimit && maxReadFileLine === -1) {
```

Comment:
I like the idea of completely circumventing this logic based on some heuristic for the current read which will minimize the surface area of this change. Does that make sense?